stl/inc/xmemory

```diff
@@ -73,27 +73,18 @@ _INLINE_VAR constexpr size_t _New_alignof = (_STD max)(alignof(_Ty),
 
 // STRUCT _Default_allocate_traits
 struct _Default_allocate_traits {
-    __declspec(allocator) static
-#ifdef __clang__ // TRANSITION, MSVC support for constexpr dynamic allocation
-        _CONSTEXPR20_DYNALLOC
-#endif // __clang__
-        void* _Allocate(const size_t _Bytes) {
+    __declspec(allocator) static _CONSTEXPR20_DYNALLOC void* _Allocate(const size_t _Bytes) {

```

Comment:
This is not correct, MSVC does no accept operator new in a constant expression, but only `construct_at`.

This is essentially a clang extension

Reply:
Note this only compiles now because `_CONSTEXPR20_DYNALLOC` gets sustituted to `inline` for MSVC

Reply:
It seems we are allowing `::operator delete` in `_Deallocate` to be invoked from constexpr context, does your comment apply to it as well?.

Reply:
That is a good question, we should ping cadacama

Reply:
Let's bring @StephanTLavavej back into the conversation, since this was his change 🐱 
To clarify, only enabling `_CONSTEXPR20_DYNALLOC` for Clang here is because MSVC and Clang have fundamentally different approaches in how they are implementing WG21-P0784: 

- Clang allows `::operator new` in `constexpr` contexts when reached transitively from a call to `std::allocator::allocate`.
- MSVC does fairy magic with `std::allocator::allocate` behind the scenes to allow it, but does not allow `::operator new` in `constexpr` contexts.  

(a bit more discussion of this in #1532)

So, I believe the conditional `_CONSTEXPR20_DYNALLOC` is not intended to be transitional - it is expected to be required even once we are able to support MSVC `constexpr` dynamic allocation.  Regarding this change, STL had [questions](https://github.com/microsoft/STL/pull/1369#discussion_r553080412) about the EDG story for WG21-P0784.  I am honestly a little fuzzy on what EDG allows/disallows in terms of `::operator new`, etc. but if it agrees with Clang in allowing it (perhaps indicated by the currently-passing tests), I think that maybe the `#ifdef` should be restored but instead of only allowing it for Clang, we should modify it to instead only _dis_-allow `_CONSTEXPR20_DYNALLOC` for MSVC.  Thoughts?

Reply:
I believe the problem is, that MSVC will find that function, because it only replaces `std::allocator::allocate` during contant evaluation. So everywhere else it will find `_Allocate` and it will throw up on it being `constexpr`

Given that we want to use MSVC for constexpr string and constexpr vector with the next release, I do not see any solution other than conditional compilation.

Reply:
Sure, I'm agreeing with the approach of conditional compilation, but am proposing to replace
```c++
#ifdef __clang__
_CONSTEXPR20_DYNALLOC
#endif // __clang__
```
with the following: 
```c++
#if !(defined(__MSVC__)) // or whatever the appropriate way to detect MSVC is
_CONSTEXPR20_DYNALLOC
#endif // __MSVC__
```
This would enable the `_Default_allocate_traits` functions to also be `constexpr` for EDG (since it appears to have no problem with it) and instead only make the functions _not_ `constexpr` for MSVC.  Then, even when we enable MSVC support for `constexpr` dynamic allocation with the next release and `_CONSTEXPR20_DYNALLOC` becomes `constexpr` for MSVC as well, these functions will not.  

If I'm understanding your concern correctly, I think this would address it.

Reply:
Yepp, totally misread what you meant :+1:

Reply:
Though now, looking at @StephanTLavavej's [comment](https://github.com/microsoft/STL/pull/1369#discussion_r553080412) more carefully (which I helpfully linked above but evidently did not fully mentally digest), he has some concerns over supporting something for EDG but not MSVC.  

> In general, it's okay to ship a feature enabled for Clang-but-not-MSVC or MSVC-but-not-Clang depending on who's ready. While EDG attempts to match MSVC, it's okay to enable a feature for MSVC but disable it for EDG temporarily (as we did with Concepts). However, customers have found the reverse to be confusing - if a feature is active for EDG but inactive for MSVC, then IntelliSense appears to be happy with the code, yet it emits errors during actual compilation (this happened for <bit>, IIRC).



So it seems like he may want to disable supporting EDG for `constexpr` dynamic allocation until we can also support it for MSVC in the next release.  I'll let him chime in more here given this discussion. 

Note that if we do this, we'll need to modify some of the tests to not expect EDG to accept `constexpr` in some places - as I recall there were some `/BE` tests that were failing before we added `defined(__EDG__)` to the `_CONSTEXPR20_DYNALLOC` macro.

Reply:
Thanks for catching my mistake!

[@mnatsuhara]
> So, I believe the conditional `_CONSTEXPR20_DYNALLOC` is not intended to be transitional - it is expected to be required even once we are able to support MSVC `constexpr` dynamic allocation.

Aha, the `TRANSITION` comment led to my confusion. 

> So it seems like he may want to disable supporting EDG for constexpr dynamic allocation until we can also support it for MSVC in the next release.

That's correct - I made that change in `yvals_core.h`, in the definition of the feature-test macro. Users are confused when IntelliSense supports features that MSVC doesn't - e.g. DevCom-997767 with `<bit>`.

When MSVC supports `constexpr` dynamic allocation, I believe it would be desirable to use the same codepaths for MSVC and EDG (as EDG IntelliSense is designed to imitate MSVC as much as possible), unless EDG rejects the MSVC approach, in which case we should (1) file an EDG bug and (2) use the Clang strategy for EDG as a workaround. (Perhaps we wouldn't even need to file an EDG bug, if this cannot possibly affect programmer-users, only the library implementation itself.)

> Note that if we do this, we'll need to modify some of the tests to not expect EDG to accept `constexpr` in some places

The tests are currently passing - I believe because they are consistently guarded with the feature-test macro. (The idea is that the feature-test macro should be the only source of truth for whether the feature is available - if it is, code should be able to use the feature without different logic for different compilers, as any such compiler workarounds should be within the product code.)

I'll restore the `#ifdef __clang__` guards to these two `_CONSTEXPR20_DYNALLOC` mentions, preserving their new location, but dropping the `TRANSITION` comments. When we enable MSVC and EDG support in the future, we can rework these guards if we find issues - the feature-test macro won't "light up" automatically, so we'll have a chance to make sure everything is good.

Reply:
Great, thanks for the explanation and fix, @StephanTLavavej!  I think you must be right - when we were hitting `/BE` failures before, it was likely because there were sections that were not correctly guarded by the feature-test macro.  I love seeing those tests green! 😸